File: Códigos Estaticos/metodosOtimizacaoPBLM.py
MAX_TAREFAS = 100
MAX_MAQUINAS = 4

class Tarefa:
    def __init__(self, tempo, indice):
        self.tempo = tempo
        self.indice = indice

def lerInstancia(arquivo):
    numeroTarefas = int(arquivo.readline().strip())

    tarefas = []
    precedencias = [[0] * MAX_TAREFAS for _ in range(MAX_TAREFAS)]

    for i in range(numeroTarefas):
        tempo = int(arquivo.readline().strip())
        tarefa = Tarefa(tempo, i + 1)
        tarefas.append(tarefa)

    while True:
        line = arquivo.readline().strip()
        if not line:
            break

        i, j = map(int, line.split(","))
        precedencias[i - 1][j - 1] = 1

    return numeroTarefas, tarefas, precedencias

def criarSolucaoInicial(numeroTarefas, tarefas, precedencias):
    maquinas = [[0] * MAX_TAREFAS for _ in range(MAX_MAQUINAS)]
    temposMaquinas = [0] * MAX_MAQUINAS

    for i in range(numeroTarefas):
        indiceTarefa = tarefas[i].indice
        maquinaAtual = min(range(MAX_MAQUINAS), key=temposMaquinas.__getitem__)

        posicaoTarefa = maquinas[maquinaAtual].index(0)
        maquinas[maquinaAtual][posicaoTarefa] = indiceTarefa
        temposMaquinas[maquinaAtual] += tarefas[i].tempo

        for j in range(i):
            if precedencias[indiceTarefa - 1][maquinas[maquinaAtual][j] - 1]:
                maquinas[maquinaAtual][posicaoTarefa] = 0
                break

    return maquinas

def calcularMakespan(numeroTarefas, tarefas, maquinas):
    temposMaquinas = [sum(tarefas[maquinas[i][j] - 1].tempo for j in range(numeroTarefas) if maquinas[i][j] > 0) for i in range(MAX_MAQUINAS)]
    makespan = max(temposMaquinas)
    maquinaMaiorMakespan = temposMaquinas.index(makespan) + 1

    print("Máquina com o maior makespan:", maquinaMaiorMakespan)

    return makespan

def criarArquivoOutput(tarefas, maquinas, makespan):
    with open("output.txt", "w") as arquivo:
        for i in range(MAX_MAQUINAS):
            arquivo.write(f"Máquina {i + 1}: ")
            arquivo.write(" ".join(str(maquinas[i][j]) for j in range(MAX_TAREFAS) if maquinas[i][j] != 0))
            arquivo.write(" (tarefas atendidas)\n")

        arquivo.write(f"FO: {makespan}\n")

    print("Arquivo de saída criado com sucesso.")

def imprimirSolucao(tarefas, maquinas):
    for i in range(MAX_MAQUINAS):
        print(f"Máquina {i + 1}:", end=" ")
        print(" ".join(str(maquinas[i][j]) for j in range(MAX_TAREFAS) if maquinas[i][j] != 0), end="")
        print(" (tarefas atendidas)")

def main():
    with open("KILBRID.IN2", "r") as arquivo:
        numeroTarefas, tarefas, precedencias = lerInstancia(arquivo)

    maquinas = criarSolucaoInicial(numeroTarefas, tarefas, precedencias)

    makespan = calcularMakespan(numeroTarefas, tarefas, maquinas)

    criarArquivoOutput(tarefas, maquinas, makespan)
    imprimirSolucao(tarefas, maquinas)
    print(f"FO: {makespan}")

if __name__ == "__main__":
    main()

File: Códigos Dinâmicos/metodosOtimizacaoPBLM.c
/* Trabalho 01 realizado para complemento de nota para Metodos de Otimização
UFES - Campus Alegre
Professor: Dayan de Castro Bissoli
Dupla: Talles Henrique 2020204803 - Alvaro Moret 
*/

#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int tempo;
    int indice;
} Tarefa;

void lerInstancia(FILE* arquivo, int* numeroTarefas, Tarefa** tarefas, int** precedencias) {
    fscanf(arquivo, "%d", numeroTarefas);

    *tarefas = (Tarefa*)malloc(*numeroTarefas * sizeof(Tarefa));
    *precedencias = (int*)calloc((*numeroTarefas) * (*numeroTarefas), sizeof(int));

    for (int i = 0; i < *numeroTarefas; i++) {
        fscanf(arquivo, "%d", &((*tarefas)[i].tempo));
        (*tarefas)[i].indice = i + 1;
    }

    int i = 0, j = 0;
    while (fscanf(arquivo, "%d,%d", &i, &j) != EOF && i != -1 && j != -1) {
        (*precedencias)[(i - 1) * (*numeroTarefas) + (j - 1)] = 1;
    }
}

void criarSolucaoInicial(int numeroTarefas, Tarefa* tarefas, int* precedencias, int numeroMaquinas, int** maquinas) {
    int* temposMaquinas = (int*)calloc(numeroMaquinas, sizeof(int));

    for (int i = 0; i < numeroTarefas; i++) {
        int indiceTarefa = tarefas[i].indice;
        int menorTempoMaquina = temposMaquinas[0];
        int maquinaAtual = 0;

        for (int j = 1; j < numeroMaquinas; j++) {
            if (temposMaquinas[j] < menorTempoMaquina) {
                menorTempoMaquina = temposMaquinas[j];
                maquinaAtual = j;
            }
        }

        int posicaoTarefa = 0;
        while ((*maquinas)[maquinaAtual * numeroTarefas + posicaoTarefa] != 0) {
            posicaoTarefa++;
        }

        (*maquinas)[maquinaAtual * numeroTarefas + posicaoTarefa] = indiceTarefa;
        temposMaquinas[maquinaAtual] += tarefas[i].tempo;

        // Verificar as precedências e ajustar a máquina atual
        for (int j = 0; j < i; j++) {
            if (precedencias[(indiceTarefa - 1) * numeroTarefas + ((*maquinas)[maquinaAtual * numeroTarefas + j] - 1)]) {
                (*maquinas)[maquinaAtual * numeroTarefas + posicaoTarefa] = 0; // Coloca a tarefa no final da máquina atual
                break;
            }
        }
    }

    free(temposMaquinas);
}

int calcularMakespan(int numeroTarefas, Tarefa* tarefas, int numeroMaquinas, int* maquinas) {
    int* temposMaquinas = (int*)calloc(numeroMaquinas, sizeof(int));

    for (int i = 0; i < numeroMaquinas; i++) {
        for (int j = 0; j < numeroTarefas; j++) {
            int indiceTarefa = maquinas[i * numeroTarefas + j];
            if (indiceTarefa > 0) {
                int tempoTarefa = tarefas[indiceTarefa - 1].tempo;
                temposMaquinas[i] += tempoTarefa;
            }
        }
    }

    int makespan = 0;
    for (int i = 0; i < numeroMaquinas; i++) {
        if (temposMaquinas[i] > makespan) {
            makespan = temposMaquinas[i];
        }
    }

    int maiorMakespan = 0;
    int maquinaMaiorMakespan = 0;

    for (int i = 0; i < numeroMaquinas; i++) {
        if (temposMaquinas[i] > maiorMakespan) {
            maiorMakespan = temposMaquinas[i];
            maquinaMaiorMakespan = i + 1;
        }
        printf("maquina :%d = %d\n", i+1, temposMaquinas[i]);
    }

    printf("Máquina com o maior makespan: %d\n", maquinaMaiorMakespan);

    free(temposMaquinas);

    return makespan;
}

void criarArquivoOutput(int numeroMaquinas, Tarefa* tarefas, int* maquinas, int makespan) {
    FILE* arquivo = fopen("output.txt", "w");
    if (arquivo == NULL) {
        printf("Erro ao criar o arquivo de saída.\n");
        return;
    }

    for (int i = 0; i < numeroMaquinas; i++) {
        fprintf(arquivo, "Máquina %d: ", i + 1);

        int j = 0;
        while (maquinas[i * numeroMaquinas + j] != 0) {
            int indiceTarefa = maquinas[i * numeroMaquinas + j];
            fprintf(arquivo, "%d ", indiceTarefa);
            j++;
        }

        fprintf(arquivo, " (tarefas atendidas)\n");
    }

    fprintf(arquivo, "FO: %d\n", makespan);

    fclose(arquivo);

    printf("Arquivo de saída criado com sucesso.\n");
}

void imprimirSolucao(int numeroMaquinas, Tarefa* tarefas, int* maquinas) {
    for (int i = 0; i < numeroMaquinas; i++) {
        printf("Máquina %d: ", i + 1);

        int j = 0;
        while (maquinas[i * numeroMaquinas + j] != 0) {
            int indiceTarefa = maquinas[i * numeroMaquinas + j];
            printf("%d ", indiceTarefa);
            j++;
        }

        printf(" (tarefas atendidas)");
        printf("\n");
    }
}

int main() {
    FILE* arquivo = fopen("KILBRID.IN2", "r");
    if (arquivo == NULL) {
        printf("Erro ao abrir o arquivo.\n");
        return 1;
    }

    int numeroTarefas;
    Tarefa* tarefas;
    int* precedencias;
    int numeroMaquinas;

    lerInstancia(arquivo, &numeroTarefas, &tarefas, &precedencias);
    fclose(arquivo);

    printf("Digite o número de máquinas: ");
    scanf("%d", &numeroMaquinas);

    int* maquinas = (int*)calloc(numeroMaquinas * numeroTarefas, sizeof(int));

    criarSolucaoInicial(numeroTarefas, tarefas, precedencias, numeroMaquinas, &maquinas);

    int makespan = calcularMakespan(numeroTarefas, tarefas, numeroMaquinas, maquinas);

    criarArquivoOutput(numeroMaquinas, tarefas, maquinas, makespan);
    imprimirSolucao(numeroMaquinas, tarefas, maquinas);
    printf("FO: %d\n", makespan);

    free(tarefas);
    free(precedencias);
    free(maquinas);

    return 0;
}